libmcount/dynamic.c

```diff
@@ -127,15 +129,25 @@ static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs)
 				stats.skipped++;
 				break;
 			case 0:
+				success = true;
 			default:
 				break;
 			}
 			stats.total++;
 		}
 
+		if (!success) {
+			nr_failed++;
+			func_name = name;
+		}
+
 		name = strtok(NULL, ";");
 	}
 
+	if (nr_failed)
+		pr_out("%s cannot be patched dynamically\n",
+		       nr_failed > 1 ? "some functions" : func_name);
+
```

Comment:
Hi @namhyung 😄  Have a good lunch!

When `stats.total` is 0, we can save first failing function name as you said.
But we cannot correctly count number of **unknown** functions (or patterns) using the if statement (i.e. if `stats.total` is 0).

For example,
### Case A
Right function name : foo
Wrong function name : wrong_name1
```
$ uftrace -P foo -P wrong_name1 ./hello.dyn
```
### The result of this Case A:
> dynamic: dynamic update stats:
> dynamic:    total:        1
> dynamic:  patched:        1 (100.00%)
> dynamic:   failed:        0 (0.00%)
> dynamic:  skipped:        0 (0.00%)

So in conclusion,
first a function name (e.g. foo) is right and it increase `stats.total` as 1,
therefore we cannot count correct number of **unknown** function name using the if statement (i.e. if `stats.total` is 0).

So what about using `stats.unknown` and `bool found` like below ?

```
diff --git a/libmcount/dynamic.c b/libmcount/dynamic.c
index f27a094..0f66f2f 100644
--- a/libmcount/dynamic.c
+++ b/libmcount/dynamic.c
@@ -17,6 +17,7 @@ static struct mcount_dynamic_stats {
         int total;
         int failed;
         int skipped;
+        int unknown;
 } stats;
 
 /* dummy functions (will be overridden by arch-specific code) */
@@ -87,7 +88,7 @@ static int prepare_dynamic_update(void)
 static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs)
 {
         char *str;
-        char *pos, *name;
+        char *pos, *name, *unknown_name = NULL;
         struct symtab *symtab = &symtabs->symtab;
 
         if (patch_funcs == NULL)
@@ -100,6 +101,7 @@ static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs)
         name = strtok(pos, ";");
         while (name) {
                 bool is_regex;
+                bool found = false;
                 regex_t re;
                 unsigned i;
                 struct sym *sym;
@@ -119,6 +121,7 @@ static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs)
                             (!is_regex && strcmp(name, sym->name)))
                                 continue;
 
+                        found = true;
                         switch (mcount_patch_func(mdinfo, sym)) {
                         case -1:
                                 stats.failed++;
@@ -133,9 +136,19 @@ static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs)
                         stats.total++;
                 }
 
+                if (!found) {
+                        unknown_name = name;
+                        stats.unknown++;
+                }
+
                 name = strtok(NULL, ";");
         }
 
+        if (!stats.failed || !stats.skipped || !stats.unknown)
+                pr_out("%s cannot be patched dynamically\n",
+                       (stats.failed + stats.skipped + stats.unknown) > 1 ?
+                       "some functions" : unknown_name);
+
         free(str);
         return 0;
 }
```

P.S.
I'm afraid that if I forcibly do `git-push` to update this PR,
`Github` remove this threads of comments.. 😟
Because this threads hang on source code line .